{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2e34818f-97d4-49e5-91d9-920d296cb606",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Creator: Kathryn Zamiela"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "61ad513f-ab2a-4ce2-952b-7347432f02d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: path to FASTA file\n",
    "# Output: DNA sequence matching the file contents\n",
    "def read_genome(filename):\n",
    "    genome = ''\n",
    "\n",
    "    with open(filename, 'r') as f:\n",
    "        for line in f:\n",
    "            # want to ignore >\n",
    "            if not line[0] == '>':\n",
    "                genome += line.rstrip() # removes trailing whitespace from end of string\n",
    "\n",
    "    return genome\n",
    "\n",
    "\n",
    "# Input: path to FASTQ file\n",
    "# Output: one list containing the DNA sequence and one list containing base quality values for the DNA sequence\n",
    "def read_fastq(filename):\n",
    "    sequences = []\n",
    "    qualities = []\n",
    "\n",
    "    with open(filename) as f:\n",
    "        while True:\n",
    "            f.readline() # ignoring first line\n",
    "            seq = f.readline().rstrip() # save sequence line\n",
    "            f.readline() # ignoring third line\n",
    "            qual = f.readline().rstrip() # save base qualities\n",
    "\n",
    "            # check if at end of file\n",
    "            if len(seq) == 0:\n",
    "                break\n",
    "\n",
    "            # save to lists\n",
    "            sequences.append(seq)\n",
    "            qualities.append(qual)\n",
    "\n",
    "    return sequences, qualities"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "1047d5bb-4b43-4af7-b0e8-ae632a6fcd52",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: string containing characters 'ACGTN'\n",
    "# Output: matching DNA base pairs in the reverse order\n",
    "def reverse_complement(s):\n",
    "    # create complement dictionary\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A', 'N': 'N'}\n",
    "\n",
    "    t = ''\n",
    "\n",
    "    for base in s:\n",
    "        t = complement[base] + t # prepending reverses the string\n",
    "\n",
    "    return t"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8e58f380-4a78-406e-ab1b-a442a0bdb392",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: string of pattern to match, string of text to match, int of position in text\n",
    "# Output: boolean indicating match status\n",
    "def is_match(pattern, text, i):\n",
    "    match = True\n",
    "\n",
    "    # check each character in the pattern based on your position in the text\n",
    "    for j in range(len(pattern)):\n",
    "     # if the characters don't match, it's not a match\n",
    "        if not text[i + j] == pattern[j]:\n",
    "            match = False\n",
    "            break\n",
    "\n",
    "    return match\n",
    "\n",
    "\n",
    "# Input: string of pattern to match, string of text to match\n",
    "# Output: a list of indices where the pattern matches the text\n",
    "#         strand-aware: will check for a match of the reverse complement also\n",
    "# Specifications: will only report a case where pattern and reverse complement match each other once\n",
    "def naive(pattern, text):\n",
    "    occurrences = [] # all indices where p matches t\n",
    "\n",
    "    # iterate through the text without going over\n",
    "    for i in range(len(text) - len(pattern) + 1):\n",
    "        # check if the pattern matches\n",
    "        match_f = is_match(pattern, text, i)\n",
    "\n",
    "        # get reverse complement of pattern\n",
    "        rev_pattern = reverse_complement(pattern)\n",
    "\n",
    "        # check if reverse complement matches\n",
    "        match_r = is_match(rev_pattern, text, i)\n",
    "        \n",
    "        # if the original pattern and reverse complement are the same, using the or statement will only save it once\n",
    "        if match_f == True or match_r == True:\n",
    "            occurrences.append(i)\n",
    "\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "ce0c1815-b101-45c4-87d0-6270097ac9f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Input: string of pattern to match, string of text to match, int of position in text\n",
    "# Output: boolean indicating match status\n",
    "# Specifications: Allows for up to 2 mismatches per occurrence\n",
    "def is_approx_match(pattern, text, i):\n",
    "    match = True\n",
    "\n",
    "    # set error counter\n",
    "    err_count = 0\n",
    "\n",
    "    # check each character in the pattern based on your position in the text\n",
    "    for j in range(len(pattern)):\n",
    "     # if the characters don't match\n",
    "        if not text[i + j] == pattern[j]:\n",
    "            # increment error counter\n",
    "            err_count += 1\n",
    "\n",
    "            # if there have been more than 2 mismatches, not a match\n",
    "            if err_count > 2:\n",
    "                match = False\n",
    "                break\n",
    "\n",
    "    return match\n",
    "    \n",
    "\n",
    "# Input: string of pattern to match, string of text to match\n",
    "# Output: a list of indices where the pattern matches the text\n",
    "# Specifications: Allows for up to 2 mismatches per occurrence\n",
    "#                 Does not consider reverse complement\n",
    "def naive_2mm(pattern, text):\n",
    "    occurrences = [] # all indices where p matches t\n",
    "\n",
    "    # iterate through the text without going over\n",
    "    for i in range(len(text) - len(pattern) + 1):\n",
    "        # check if the pattern matches\n",
    "        match = is_approx_match(pattern, text, i)\n",
    "        \n",
    "        # if the original pattern and reverse complement are the same, using the or statement will only save it once\n",
    "        if match == True:\n",
    "            occurrences.append(i)\n",
    "\n",
    "    return occurrences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "ddfea992-222d-4289-b329-d0353a026160",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Sample usage of this block: \n",
    "# seqs, quals = read_fastq(filename)\n",
    "# worst_idx = find_worst_seq_cycle(quals)\n",
    "# print(worst_idx)\n",
    "\n",
    "\n",
    "# Input: Phred+33 int value\n",
    "# Output: character representing base quality from FASTQ file\n",
    "def q_to_phred33(q):\n",
    "    return chr(q + 33) # chr converts char to int according to ASCII\n",
    "\n",
    "\n",
    "# Input: character representing base quality from FASTQ file\n",
    "# Output: Phred+33 int value\n",
    "def phred33_to_q(qual):\n",
    "    return ord(qual)-33 # ord converts int to char according to ASCII\n",
    "\n",
    "\n",
    "# Input: list of base qualities from FASTQ file\n",
    "# Output: list of average qualities, index corresponds to FASTQ read index\n",
    "# Specifications: assumes that all reads are of the same length\n",
    "def create_seq_cycle_avgs(quals):\n",
    "    seq_cycle_avg = []\n",
    "\n",
    "    # calculate length of read\n",
    "    len_read = len(quals[0])\n",
    "\n",
    "    # iterate through each base\n",
    "    for i in range(len_read):\n",
    "        # initialize total\n",
    "        total_qual = 0\n",
    "\n",
    "        # iterate through each read\n",
    "        for qual in quals:\n",
    "            # get int value\n",
    "            val = phred33_to_q(qual[i])\n",
    "\n",
    "            # add to total\n",
    "            total_qual += val\n",
    "\n",
    "        # calculate average\n",
    "        seq_cycle_avg.append(total_qual / float(len_read))\n",
    "\n",
    "    # return list of averages\n",
    "    return seq_cycle_avg\n",
    "\n",
    "\n",
    "# Input: list of sequencing cycle averages from FASTQ data\n",
    "# Output: a float of the lowest average found as well as index found at\n",
    "def find_lowest_val_with_idx(l):\n",
    "    lowest_val = 0.0\n",
    "    lowest_idx = -1\n",
    "\n",
    "    for i in range(len(l)):\n",
    "        # if lowest val hasn't been set, use current val\n",
    "        if lowest_idx == -1:\n",
    "            lowest_val = l[i]\n",
    "            lowest_idx = i\n",
    "        elif l[i] < lowest_val:\n",
    "            lowest_val = l[i]\n",
    "            lowest_idx = i\n",
    "\n",
    "    return lowest_val, lowest_idx\n",
    "\n",
    "\n",
    "# Input: list of base qualities from FASTQ file\n",
    "# Output: index of poorest sequencing cycle\n",
    "# Specifications: assumes that all reads are of the same length\n",
    "def find_worst_seq_cycle(quals):\n",
    "    # create list of cycle averages\n",
    "    seq_cycle_avg = create_seq_cycle_avgs(quals)\n",
    "\n",
    "    # find lowest average\n",
    "    lv, li = find_lowest_val_with_idx(seq_cycle_avg)\n",
    "    \n",
    "    return li"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c845cd87-e16a-4fb2-9cbc-8e3e8667f83b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
